---
title: "Post-alignment QC workflow"
output: html_document
fontsize: 11pt
geometry: margin=1in
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Tools

- RSeQC - http://rseqc.sourceforge.net/#reference - calculates different metrics specialised for RNA-seq data
- 

### converting gtf to bed format

```{r gtf to bed, eval = F, echo = T}
/tools/ea-utils/ExpressionAnalysis-ea-utils-bd148d4/clipper/gtf2bed \
/data/references/ensembl/gtf_gff3/v95/Homo_sapiens.GRCh38.95.gtf \
> /data/references/ensembl/bed/v95/ensembl_GRCh38_v95.bed
```

### Workflow

- Currently, the modules chosen are: 
    - geneBody_coverage.py
    - clipping_profile.py
    - inner_distance.py
    - junction_annotation.py
    - junction_saturation.py
    - mismatch_profile.py - check 
    - read_distribution.py
    - read_duplication.py
    - read_GC.py
    - RNA_fragment_size.py
    - tin.py
    - bam_stat.py
    



